torch/distributed/device_mesh.py

```diff
@@ -488,6 +325,162 @@ def __init__(
+            if group is not None:
+                layouts_to_groups_map[layout] = group.group_name
+                return
+
+            default_group = _get_default_group()
+
+            if layout == _Layout(((get_world_size(), 1),)) and backend_override == (
+                None,
+                None,
+            ):
+                # Append the default pg to the first dim groups only if the default pg is compatible with `self.device_type`.
+                # Otherwise, create new pg.
+                ranks = list(range(get_world_size()))
+                group = (
+                    new_group(
+                        backend="cpu:gloo,cuda:nccl",
+                        ranks=ranks,
+                        group_desc="mesh_default",
+                    )
+                    if torch.cuda.is_available()
+                    and get_backend(default_group) == "gloo"
+                    else default_group
+                )
+            else:
+                # Generate the pg_ranks_by_dim for pg_creation from the layout.
+                pg_ranks_by_dim = array_mesh if array_mesh is not None else layout.layout_to_global_ranks(get_world_size())
+                backend, pg_options = backend_override
+
+                # If we have a 2D mesh with mesh_dim_names ("dp", "tp"), the group description
+                # of the subgroups would be `mesh_dim_dp` and `mesh_name_tp`.
+                # If the mesh doesn't not have a mesh_dim_names, then the group description of the
+                # subgroup would be `mesh_dim_0` and `mesh_dim_1`.
+                group_desc = f"mesh_{name}" if name is not None else f"mesh_dim_{dim}"
+
+                # If bound_device_id exists, it means the nccl communicator has been eagerly initialized
+                # so that we can use `split_group` to create subgroups through `ncclCommSplit`.
+                # In this case, we only need to make one API call (`split_group``) for the subgroup creation
+                # for each mesh dimension. In a 2 * 4 mesh, we only need to make 2 API calls per ranks to create
+                # all the subgroups.
+                # Otherwise, we need to make more than one API call (`new_group`) for subgroup creations. The
+                # numbers of API calls are equal to the number of subgroups for each mesh dimension. In a 2 * 4
+                # mesh, we need to make 2 + 4 = 6 API calls per ranks to create all the subgroups.
+                group = None
+                if (
+                    getattr(default_group, "bound_device_id", None) is not None
+                    and torch.cuda.is_available()
+                    and (
+                        backend is None
+                        or default_group._get_backend(torch.device("cuda")).name()
+                        == backend
+                    )
+                ):
+                    group = split_group(
+                        parent_pg=default_group,
```

Comment:
this seems to imply that we're only allowed to perform splits off of the default group.  is there a reason for this? 
(this can be something to address in the future, i suspect this was pre-existing behavior)

Reply:
you misunderstood me. 

If i create a mesh with (DP, TP) and then later I split DP into (DP_shard, DP_replicate)- should I be allowed to use comm_split for this split?  Does it depend on whether DP group was created with new group in the first place, or split from default group in the first place?

Reply:
this is a great point, let me add a TODO here, I literally just copied the existing code here. (https://github.com/pytorch/pytorch/blob/3bfa35d62e7c5df2b651a3d19b47db74ef284788/torch/distributed/device_mesh.py#L539C13-L539C26)